### PR TITLE
maybe, either: simplify S.encase and S.encaseEither

### LINKED ISSUE #103: simplify S.encase and S.encaseEither
Currently, [`encase`][1] and [`encaseEither`][2] have misleading type signatures:

```haskell
encase :: (* -> a) -> (* -> Maybe a)

encaseEither :: (Error -> a) -> (* -> b) -> (* -> Either a b)
```

They are misleading because Hindley–Milner notation cannot represent a function which returns a curried function of unknown arity.

In addition to the documentation problem, there's a usability problem. Functions with unwanted optional arguments must be wrapped in lambdas. For example:

```javascript
S.encaseEither(S.I, x => JSON.parse(x))
```

A solution to both problems is to have one function for each arity up to some arbitrary limit:

```haskell
encase  :: (a -> b) -> a -> Maybe b
encase2 :: (a -> b -> c) -> a -> b -> Maybe c
encase3 :: (a -> b -> c -> d) -> a -> b -> c -> Maybe d

encaseEither  :: (Error -> l) -> (a -> r) -> a -> Either l r
encaseEither2 :: (Error -> l) -> (a -> b -> r) -> a -> b -> Either l r
encaseEither3 :: (Error -> l) -> (a -> b -> c -> r) -> a -> b -> c -> Either l r
```

An incidental benefit of this approach is that arguments to the wrapped function may be provided up front. For example:

```javascript
> S.encaseEither(R.prop('message'), JSON.parse, '[1,2,3]')
Right([1, 2, 3])
> S.encaseEither(R.prop('message'), JSON.parse, 'XXX')
Left("Unexpected token X")
```

[1]: https://github.com/plaid/sanctuary#encase
[2]: https://github.com/plaid/sanctuary#encaseEither

### PR DESCRIPTION
**Pull request by davidchambers**
Closes #103

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -2016,9 +2058,7 @@
   //. > S.parseJson('[')
   //. Nothing()
   //. ```
-  S.parseJson = def('parseJson', [String], encase(function(s) {
-    return JSON.parse(s);
-  }));
+  S.parseJson = def('parseJson', [String], encase(JSON.parse));
```
Comment (CrisFeo) [inline index.js]: Nice!

**Comment by CrisFeo** [comment]
Funny how sometimes simplifying requires adding 200 more lines ;)

Good work on this @davidchambers! :moyai:

File: index.js (hunk 1 above)
Comment (davidchambers) [inline index.js]: It'll be nice to no longer be dependent on the `length` of the provided function. :)

**Comment by davidchambers** [comment]
Are you in favour of these changes, @svozza?

**Comment by svozza** [comment]
Initially I was bit hesitant about but having seen how fixing arity simplifies the implementations of functions (not just in this PR) I'm fully supportive of this direction now.

**Pull request merged**